crypto/ec/curve448/arch_32/f_impl.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2014 Cryptography Research, Inc.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ * Originally written by Mike Hamburg
+ */
+
+#include "field.h"
+
+#if (defined(__OPTIMIZE__) && !defined(__OPTIMIZE_SIZE__) && !defined(I_HATE_UNROLLED_LOOPS)) \
+     || defined(C448_FORCE_UNROLL)
+# define REPEAT8(_x) _x _x _x _x _x _x _x _x
+# define FOR_LIMB(_i,_start,_end,_x) do { _i=_start; REPEAT8( if (_i<_end) { _x; } _i++;) } while (0)
+#else
+# define FOR_LIMB(_i,_start,_end,_x) do { for (_i=_start; _i<_end; _i++) _x; } while (0)
+#endif
+
+void gf_mul(gf_s * RESTRICT cs, const gf as, const gf bs)
+{
+    const uint32_t *a = as->limb, *b = bs->limb;
+    uint32_t *c = cs->limb;
+    uint64_t accum0 = 0, accum1 = 0, accum2 = 0;
+    uint32_t mask = (1 << 28) - 1;
+    uint32_t aa[8], bb[8];
+    int i, j;
+
+    for (i = 0; i < 8; i++) {
+        aa[i] = a[i] + a[i + 8];
+        bb[i] = b[i] + b[i + 8];
+    }
+
+    FOR_LIMB(j, 0, 8, {
+             accum2 = 0;
+             FOR_LIMB(i, 0, j + 1, {
+                      accum2 += widemul(a[j - i], b[i]);
+                      accum1 += widemul(aa[j - i], bb[i]);
+                      accum0 += widemul(a[8 + j - i], b[8 + i]);
+                      }
+             ); accum1 -= accum2; accum0 += accum2;
+             accum2 = 0;
+             FOR_LIMB(i, j + 1, 8, {
+                      accum0 -=
+                      widemul(a[8 + j - i], b[i]);
+                      accum2 +=
+                      widemul(aa[8 + j - i],
+                              bb[i]);
+                      accum1 += widemul(a[16 + j - i], b[8 + i]);
+                      }
+             );
+             accum1 += accum2;
+             accum0 += accum2;
+             c[j] = ((uint32_t)(accum0)) & mask;
+             c[j + 8] = ((uint32_t)(accum1)) & mask;
+             accum0 >>= 28; accum1 >>= 28;
+             });
+
+    accum0 += accum1;
+    accum0 += c[8];
+    accum1 += c[0];
+    c[8] = ((uint32_t)(accum0)) & mask;
+    c[0] = ((uint32_t)(accum1)) & mask;
+
+    accum0 >>= 28;
+    accum1 >>= 28;
+    c[9] += ((uint32_t)(accum0));
+    c[1] += ((uint32_t)(accum1));
+}
+
+void gf_mulw_unsigned(gf_s * RESTRICT cs, const gf as, uint32_t b)
+{
+    const uint32_t *a = as->limb;
+    uint32_t *c = cs->limb;
+    uint64_t accum0 = 0, accum8 = 0;
+    uint32_t mask = (1 << 28) - 1;
+    int i;
+
+    assert(b < 1 << 28);
+
+    FOR_LIMB(i, 0, 8, {
```

Comment:
I'm not sure I like this new for loop syntax...

Reply:
It is as written in the original implementation by Mike Hamburg. We could choose to rewrite it anyway we like. Do you have any ideas about how else to do it? One of the problems of putting code blocks in a macro like this is that it makes it difficult to run through a debugger.

Reply:
Yes, I just saw the ugliness so far, but debugging must be near impossible this way.
I think this "FOR_LIMB(i, 0, 8, {" does not add much value over "for (i=0; i<8; i++) {"

Reply:
Well I think the justification for it is performance, i.e. we normally use this version of the macro:

````
# define REPEAT8(_x) _x _x _x _x _x _x _x _x
# define FOR_LIMB(_i,_start,_end,_x) do { _i=_start; REPEAT8( if (_i<_end) { _x; } _i++;) } while (0)
````

Which is written specifically to *avoid* a loop. The original code has been significantly optimised by Mike. During importing the code I was trying to avoid doing anything that was going to break that optimisation.

Reply:
Hmm....

This looks more like a duff's device

`# define FOR_LIMB(_i, _start, _end, _x) do { (_i) = (_start); switch((_end)-(_start)) { case 8: _x; (_i)++; case 7:  _x;  (_i)++;  .... case 1: _x; (_i)++; } } while(0)`

Reply:
I used your version of the macro.